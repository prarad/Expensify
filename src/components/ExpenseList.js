import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import getVisibleExpenses from '../selectors/expenses'

export const ExpenseList = props => (
    <div>
        <h1>Expense List</h1>
        {props.expenses.map(expense => (
            <ExpenseListItem {...expense} key={expense.id} />
        ))}
    </div>
)
const mapStateToProps = state => ({
    expenses: getVisibleExpenses(state.expenses, state.filters)
})

export default connect(mapStateToProps)(ExpenseList)