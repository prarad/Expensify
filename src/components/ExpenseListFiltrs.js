import React from 'react';
import uuid from 'uuid'
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import { setTextFilter, sortByDate, sortByAmount, setEndDate, setStartDate } from '../actions/filters';

export class ExpenseListFilters extends React.Component {
  state = {
    calenderFocused: null
  }
  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }
  onFocusChange = calenderFocused => {
    this.setState(() => ({ calenderFocused }))
  }
  onTextChange = e => {
    this.props.setTextFilter(e.target.value)
  }
  onSortChange = e => {
    e.target.value === 'date' ?
      this.props.sortByDate() :
      this.props.sortByAmount()
  }
  render() {
    return (
      <div className="ExpenseListFilters">
        <div className="textFilter">
          <label htmlFor="text-filter">Filter By Text: </label>
          <input
            id="text-filter"
            type="text"
            value={this.props.filters.text}
            onChange={this.onTextChange} />
        </div>
        <div className="sort-by-filter">
          <label htmlFor="sort-by-filter">Sort By: </label>
          <select
            name="sort-by"
            id="sort-by-filter"
            value={this.props.filters.sortBy}
            onChange={this.onSortChange}
          >
            <option value="date">Date</option>
            <option value="amount">Amount</option>
          </select>
          {this.props.filters.sortBy === 'date' &&
            (
              <DateRangePicker
                id='date-range-picker'
                startDate={this.props.filters.startDate}
                startDateId='start-date-id'
                endDate={this.props.filters.endDate}
                endDateId='end-date-id'
                onDatesChange={this.onDatesChange}
                focusedInput={this.state.calenderFocused}
                onFocusChange={this.onFocusChange}
                showClearDates={true}
                numberOfMonths={1}
                isOutsideRange={() => false}
              />
            )
          }
        </div>
      </div>
    )
  }
}
const mapStateToProps = state => ({ filters: state.filters })
const mapDispatchToProps = dispatch => ({
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate)),
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount())
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)