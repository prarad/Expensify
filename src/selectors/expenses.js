import moment from 'moment';

export default (expenses, {
  text,
  sortBy,
  startDate,
  endDate
}) => {
  return expenses.filter(expense => {
    const createdAtMoment = moment(expense.createdAt);
    const words = expense.description.split(' ')
      .map(word => word.toLowerCase());
    const textMatch = !text || words.includes(text.toLowerCase());
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;

    return textMatch && startDateMatch && endDateMatch
  }).sort((a, b) => (
    sortBy === 'date' ?
      a.createdAt < b.createdAt :
      a.amount < b.amount
  ))
}