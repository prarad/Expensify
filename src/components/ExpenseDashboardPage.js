import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFiltrs';

export default () => (
  <div className="ExpenseDashboardPage">
    <ExpenseListFilters />
    <ExpenseList />
  </div>
)