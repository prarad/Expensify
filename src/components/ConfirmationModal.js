import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import { closeModal } from '../actions/confirmationModal';
import { startRemoveExpense } from '../actions/expenses';

export const ConfirmationModal = ({
  modalIsOpen,
  handleRequestClose,
  startRemoveExpense,
  selectedExpense = {},
  history
}) => (
    <Modal
      className="modal"
      isOpen={modalIsOpen}
      contentLabel="Confirmation Modal"
      onRequestClose={handleRequestClose}
      closeTime={200}
      ariaHideApp={false}
    >
      <h2 className="modal__title">Confirmation</h2>
      <p className="modal__body">
        Are you sure to delete '{<span>{selectedExpense.description}</span>}' item?
    </p>
      <div className="modal__answers">
        <button
          className="btn"
          onClick={() => {
            startRemoveExpense(selectedExpense.id)
              .then(() => {
                handleRequestClose();
                history.push('/');
              })
          }}
        >Yes</button>
        <button className="btn" onClick={handleRequestClose}>No</button>
      </div>
    </Modal>
  );

const mapStateToProps = state => ({
  modalIsOpen: state.ModalIsOpen
});

const mapDispatchToProps = dispatch => ({
  handleRequestClose: () => dispatch(closeModal()),
  startRemoveExpense: id => dispatch(startRemoveExpense(id))
})

export default connect(mapStateToProps, mapDispatchToProps)(ConfirmationModal);