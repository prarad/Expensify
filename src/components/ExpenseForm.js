import React from 'react';
import { addExpense } from '../actions/expenses'
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';
import 'react-dates/lib/css/_datepicker.css'

export default class ExpenseForm extends React.Component {
  state = {
    description: '',
    amount: '',
    note: '',
    createdAt: moment(),
    calendarFocused: false,
    error: ''
  }
  dispatch = this.props.dispatch;

  onDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  }
  onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d+(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }))
    }
  }
  onNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }))
  }
  onDateChange = createdAt => {
    createdAt &&
      this.setState(() => ({ createdAt }))
  }
  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }))
  }
  onSubmit = e => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Please provide description and amount'
      }))
    } else {
      this.setState(() => ({ error: '' }))
    }
    !this.state.error && this.dispatch(
      addExpense({
        description: this.state.description,
        note: this.state.note,
        amount: this.state.amount,
        createdAt: this.state.createdAt
      })
    )
  }
  render() {
    return (
      <div className='Expense-Form'>

        {!!this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}>
          <input type="text"
            placeholder='Description'
            onChange={this.onDescriptionChange}
            autoFocus
          />
          <input type="text"
            value={this.state.amount}
            onChange={this.onAmountChange}
            placeholder='Amount'
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea name="" id="" cols="30" rows="10"
            placeholder='Add a note for your expense(optional)'
            onChange={this.onNoteChange}
          ></textarea>
          <button>Add Expense</button>
        </form>
      </div >
    )
  }
}