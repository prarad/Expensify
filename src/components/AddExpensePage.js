import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { addExpense } from '../actions/expenses';

const AddExpensePage = ({ dispatch, history }) => (
  <div className='AddExpensePage'>
    <h1>Add Expense</h1>
    <ExpenseForm
      action='add'
      onSubmit={(expense => {
        dispatch(addExpense(expense))
        history.push('/')
      })} />
  </div>
)

export default connect()(AddExpensePage)