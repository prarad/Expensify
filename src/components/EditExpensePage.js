import React from 'react';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.editExpense({
      id: this.props.expense.id,
      updates: expense
    });
    this.props.history.push('/')
  }
  onRemove = () => {
    this.props.removeExpense(this.props.expense.id)
    this.props.history.push('/')
  }
  render() {
    return ( < div className='AddExpensePage' >
        <h1>Edit Expense</h1>
        <ExpenseForm
          expense={this.props.expense}
          onSubmit={this.onSubmit}
        />
        <button
          onClick={this.onRemove}
        >Remove</button>
      </div >
    )
  }
}

const mapDispatchToProps = (dispatch, props) => ({
  removeExpense: id => dispatch(removeExpense(id)),
  editExpense: (id, expense) => dispatch(editExpense({ id, updates: expense }))
})

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.id === props.match.params.id)
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)