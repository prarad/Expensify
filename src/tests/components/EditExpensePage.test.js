import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from '../../components/EditExpensePage';
import expenses from '../fixtures/expenses';

let wrapper, removeExpense, editExpense, history, expense = expenses[2];

beforeEach(() => {
  editExpense = jest.fn();
  removeExpense = jest.fn();
  history = { push: jest.fn() }
  wrapper = shallow(<EditExpensePage
    removeExpense={removeExpense}
    editExpense={editExpense}
    history={history}
    expense={expense}
  />
  );
});

test('should render EditExpensePage correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test('should handle edit expense correctly', () => {
  wrapper.find('ExpenseForm').prop('onSubmit')(expense);
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(editExpense).toHaveBeenLastCalledWith({ id: expense.id, updates: expense })
});

test('should handle remove expense correctly', () => {
  wrapper.find('button').simulate('click');
  expect(history.push).toHaveBeenLastCalledWith('/');
  expect(removeExpense).toHaveBeenLastCalledWith(expense.id);
});