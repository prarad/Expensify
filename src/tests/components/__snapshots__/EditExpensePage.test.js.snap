// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div
  className="AddExpensePage"
>
  <h1>
    Edit Expense
  </h1>
  <ExpenseForm
    expense={
      Object {
        "amount": 141.35,
        "createdAt": 345600000,
        "description": "Coffee and dinner",
        "id": "3",
        "note": "",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
